src/storage/single_file_block_manager.cpp

```diff
@@ -114,11 +122,11 @@ void MainHeader::Write(WriteStream &ser) {
 	SerializeVersionNumber(ser, DuckDB::LibraryVersion());
 	SerializeVersionNumber(ser, DuckDB::SourceID());
 
-	if (flags[0] == MainHeader::ENCRYPTED_DATABASE_FLAG) {
-		SerializeEncryptionMetadata(ser, encryption_metadata);
-		SerializeSalt(ser, salt);
-		SerializeEncryptionMetadata(ser, encrypted_canary);
-	}
+	// We always serialize, and write zeros, if not set.
+	auto encryption_enabled = flags[0] & MainHeader::ENCRYPTED_DATABASE_FLAG;
+	SerializeEncryptionMetadata(ser, encryption_metadata, encryption_enabled);
+	SerializeDBIdentifier(ser, db_identifier);
+	SerializeEncryptionMetadata(ser, encrypted_canary, encryption_enabled);
```

Comment:
Coming back on this, I think given we are touching this we should have a check like:
```
if (flag[0] > max_flag_number_for_my_db_version) {
  throw;
}
```
both here and on the writing side.
Currenlty we do rely on 0 always being there for versions < 1.3.0, and 0 or 1 being there for versions <= 1.4.0, basically this being tied to duckdb version implicitly, and us doing no mistakes in the future, but I think it's more appropriate to encode that in the code (follow up)

Reply:
Then cool, thanks